utils.go

```diff
@@ -73,3 +75,42 @@ func contains(stringList []string, element string) bool {
 	}
 	return false
 }
+
+// shutdownSignal - is the channel that receives any boolean when
+// we want broadcast the start of shutdown
+var shutdownSignal chan bool
+
+// shutdownCallbacks - is the list of function callbacks executed one by one
+// when a shutdown starts. A callback returns 0 for success and 1 for failure.
+// Failure is considered an emergency error that needs an immediate exit
+var shutdownCallbacks []func() int
+
+// Register callback functions that needs to be called when process terminates.
+func registerShutdown(callback func() int) {
+	shutdownCallbacks = append(shutdownCallbacks, callback)
+}
+
+// Start to monitor shutdownSignal to execute shutdown callbacks
+func monitorShutdownSignal() {
+	go func() {
+		// Monitor processus signal
+		trapCh := signalTrap(os.Interrupt, syscall.SIGTERM)
+		for {
+			select {
+			case <-trapCh:
+				// Start a graceful shutdown call
+				shutdownSignal <- true
+			case <-shutdownSignal:
+				// Call all callbacks and exit for emergency
+				for _, callback := range shutdownCallbacks {
+					i := callback()
```

Comment:
yes, it is just because shutdownFS() was exiting using a return code and I didn't want to break things.. shall we rename `i` to `exitCode` or we use `nil` and `error` instead ?